test/k8sT/Egress.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+	assignEgressIP := func() {
+		// Assign egress IP address to k8s2
+		originalAssignIPYAML := helpers.ManifestGet(kubectl.BasePath(), "egress-ip-deployment.yaml")
+		res := kubectl.ExecMiddle("mktemp")
+		res.ExpectSuccess()
+		assignIPYAML = strings.Trim(res.Stdout(), "\n")
+		kubectl.ExecMiddle(fmt.Sprintf("sed 's/INPUT_EGRESS_IP/%s/' %s > %s",
+			egressIP, originalAssignIPYAML, assignIPYAML)).ExpectSuccess()
+		res = kubectl.ApplyDefault(assignIPYAML)
+		Expect(res).Should(helpers.CMDSuccess(), "unable to apply %s", assignIPYAML)
+	}
+
+	BeforeAll(func() {
+		if !helpers.ExistNodeWithoutCilium() {
+			Skip("EgressGatewayTest requires outside nodes")
+		}
+
+		kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
+
+		_, k8s1IP = kubectl.GetNodeInfo(helpers.K8s1)
+		_, k8s2IP = kubectl.GetNodeInfo(helpers.K8s2)
+		_, outsideIP = kubectl.GetNodeInfo(helpers.GetNodeWithoutCilium())
+
+		egressIP = getEgressIP(k8s1IP)
+
+		deploymentManager.SetKubectl(kubectl)
+
+		ciliumFilename = helpers.TimestampFilename("cilium.yaml")
+		DeployCiliumAndDNS(kubectl, ciliumFilename)
+
+		runEchoServer()
+		assignEgressIP()
+	})
+
+	AfterAll(func() {
+		_ = kubectl.Delete(echoPodYAML)
+		_ = kubectl.Delete(assignIPYAML)
+
+		UninstallCiliumFromManifest(kubectl, ciliumFilename)
+	})
+
+	AfterFailed(func() {
+		kubectl.CiliumReport("cilium bpf egress list", "cilium endpoint list")
+	})
+
+	testEgressGateway := func(fromGateway bool) {
+		hostIP := k8s1IP
+		if fromGateway {
+			hostIP = k8s2IP
+		}
+		srcPod, _ := fetchPodsWithOffset(kubectl, randomNamespace, "client", "zgroup=testDSClient", hostIP, false, 1)
+
+		res := kubectl.ExecPodCmd(randomNamespace, srcPod, helpers.CurlFail("http://%s:80", outsideIP))
+		res.ExpectSuccess()
+		res.ExpectMatchesRegexp(fmt.Sprintf("client_address=::ffff:%s\n", egressIP))
+	}
+
+	applyEgressPolicy := func() {
+		// Apply egress policy yaml
+		originalPolicyYAML := helpers.ManifestGet(kubectl.BasePath(), "egress-policy-demo.yaml")
+		res := kubectl.ExecMiddle("mktemp")
+		res.ExpectSuccess()
+		policyYAML = strings.Trim(res.Stdout(), "\n")
+		kubectl.ExecMiddle(fmt.Sprintf("sed 's/INPUT_EGRESS_IP/%s/' %s > %s",
+			egressIP, originalPolicyYAML, policyYAML)).ExpectSuccess()
+		kubectl.ExecMiddle(fmt.Sprintf("sed 's/INPUT_OUTSIDE_NODE_IP/%s/' -i %s",
+			outsideIP, policyYAML)).ExpectSuccess()
+		res = kubectl.ApplyDefault(policyYAML)
+		Expect(res).Should(helpers.CMDSuccess(), "unable to apply %s", policyYAML)
+	}
+
+	Context("tunnel disabled", func() {
+		BeforeAll(func() {
+			DeployCiliumOptionsAndDNS(kubectl, ciliumFilename, map[string]string{
+				"egressGateway.enabled": "true",
+				"tunnel":                "disabled",
+				"bpf.masquerade":        "true",
+			})
+
+			randomNamespace = deploymentManager.DeployRandomNamespaceShared(DemoDaemonSet)
+			deploymentManager.WaitUntilReady()
+
+			applyEgressPolicy()
+			// There is no easy way to wait for policy to be reconciled now
+			time.Sleep(time.Second)
```

Comment:
Could we check the BPF map content? It's clearly not ideal, but I'm worried a sleep in the code will cause a lot of flakes.